In [2]:
import pandas as pd

# 파일 수집

In [3]:
file_path = 'data/boston.csv'

In [4]:
bost = pd.read_csv(file_path)
print(bost.columns)

Index(['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax',
       'ptratio', 'b', 'lstat', 'medv'],
      dtype='object')


In [5]:
bost.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [6]:
bost.tail()

crim   zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
501  0.06263  0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273     21.0   
502  0.04527  0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273     21.0   
503  0.06076  0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273     21.0   
504  0.10959  0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273     21.0   
505  0.04741  0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273     21.0   

          b  lstat  medv  
501  391.99   9.67  22.4  
502  396.90   9.08  20.6  
503  396.90   5.64  23.9  
504  393.45   6.48  22.0  
505  396.90   7.88  11.9

In [7]:
# 독립 = independent, 종속 = dependent
# 속성은 대소문자 구분하지 않음
indepen = bost[['crim', 'zn', 'indus', 'chas', 'nox', 'rm', 'age', 'dis', 'rad', 'tax', 'ptratio', 'b', 'lstat']]
depen = bost[['medv']]
print(indepen.shape, depen.shape)

(506, 13) (506, 1)


In [8]:
# 모델의 구조 새성
import tensorflow as tf
x = tf.keras.layers.Input(shape=[13])
y = tf.keras.layers.Dense(1)(x)
model = tf.keras.models.Model(x, y)
model.compile(loss='mse')

In [9]:
# 모델 학습
model.fit(indepen, depen, epochs=1000)

Train on 506 samples
Epoch 1/1000
506/506 [==============================] - 84s 166ms/sample - loss: 149062.5582
Epoch 2/1000
506/506 [==============================] - 0s 77us/sample - loss: 136146.4830
Epoch 3/1000
506/506 [==============================] - 0s 63us/sample - loss: 125518.7050
Epoch 4/1000
506/506 [==============================] - 0s 78us/sample - loss: 115507.7271
Epoch 5/1000
506/506 [==============================] - 0s 68us/sample - loss: 105906.3032
Epoch 6/1000
506/506 [==============================] - 0s 59us/sample - loss: 96720.5095
Epoch 7/1000
506/506 [==============================] - 0s 81us/sample - loss: 88038.4855
Epoch 8/1000
506/506 [==============================] - 0s 56us/sample - loss: 79728.9507
Epoch 9/1000
506/506 [==============================] - 0s 81us/sample - loss: 71950.7964
Epoch 10/1000
506/506 [==============================] - 0s 75us/sample - loss: 64544.1111
Epoch 11/1000
506/506 [==============================] - 0s 67us/sample

In [11]:
# 모델 이용
model.predict(indepen[0:5])

array([[28.816013],
       [24.397194],
       [29.97523 ],
       [29.31436 ],
       [28.682968]], dtype=float32)

In [12]:
# 실제 csv 값의 종속값
depen[0:5]

medv
0  24.0
1  21.6
2  34.7
3  33.4
4  36.2

In [13]:
# 모델 수식 확인
model.get_weights()

[array([[-0.0886721 ],
        [ 0.07566881],
        [-0.05553464],
        [ 3.326849  ],
        [ 0.97685677],
        [ 3.897722  ],
        [ 0.01877917],
        [-0.88825774],
        [ 0.14266615],
        [-0.00972757],
        [ 0.06900827],
        [ 0.01590026],
        [-0.58691907]], dtype=float32),
 array([2.131235], dtype=float32)]

# 모델 수식
bhp = [-0.0886721 ] * CRIM +
        [ 0.04774402] * ZN +
        [-0.0208921 ] * INDUS +
        [ 2.5016797 ] * CHAS +
        [-2.0003839 ] * NOX +
        [ 5.3016825 ] * RM +
        [-0.00753869] * AGE +
        [-1.0406284 ] * DIS +
        [ 0.17047067] * RAD +
        [-0.00949569] * TAX +
        [-0.39329278] * PTRATIO +
        [ 0.01451861] * B +
        [-0.41769162] * LSTAT +
        [ 0.]